Use new `memory_stats` call; use pandas for formatting

andyljones opened the pull request:
This PR
  * Adds [detailed memory stats](https://pytorch.org/docs/stable/cuda.html#torch.cuda.memory_stats) 
  * Handles peaks using `reset_peak_memory_stats()` rather than emptying the cache 
  * Dumps the trace as a pandas dataframe, which can be used for downstream analysis
  * Does the formatting using pandas' `to_string`, so variable numbers of fields are supported

Upshot is that `.records()` will give you a dataframe like this 

![image](https://user-images.githubusercontent.com/662606/79572593-1e0fa880-80b5-11ea-9ed2-4548b6fa4e97.png)

with 96 columns covering every memory stat Torch offers. Terminal output can be whatever subset you want:

![image](https://user-images.githubusercontent.com/662606/79573060-efde9880-80b5-11ea-9b41-2d7e6a54d5eb.png)

I don't think you'll want to take this PR as is; I've broken a bunch of stuff getting to what I need for the task at hand. I'm more making this PR in case you want to nab parts for the main branch.

andyljones commented:
now with bonus HTML barcharts

![image](https://user-images.githubusercontent.com/662606/79578413-1accea80-80be-11ea-9601-88b3f0055001.png)

andyljones commented:
And now it accurately reports peaks and allocated/freed when you're profiling a hierarchy of functions. 

![image](https://user-images.githubusercontent.com/662606/79632915-3259b100-815a-11ea-879f-97d87f744d59.png)

This has been _really_ useful for me, thanks so much for making it!

I've tried to restore some of the things I know that I broke, but some other stuff is still going to block merging with main. In particular,

  * It only does HTML output right now, no text
  * It requires pandas as a dependency. This might be too heavy for your tastes.

Stonesjtu commented:
It's great that you liked this tool. This is really a big PR, hopefully I can review these changes thoroughly this weekend.

andyljones commented:
Thanks! No need to feel rushed though, and I understand if you decide it's too far from the original tool to be justifiable to fold in.

Stonesjtu commented on pytorch_memlab/line_profiler.py:
```diff
@@ -58,124 +60,165 @@ def add_function(self, func):
+        template = '<h3><span style="font-family: monospace">{q}</span></h3><div>{c}</div>'
+        html = '\n'.join(template.format(q=q, c=c) for q, c in chunks)
+        if stream is None:
+            display(HTML(html))
```
Will this support printing to the command line interface?

andyljones replied:
No. Let's discuss back in the main thread what to do about the command line.

Stonesjtu commented on pytorch_memlab/line_profiler.py:
```diff
@@ -207,65 +250,40 @@ def forward(self, inp):
 
     return func
 
+def profile_every(output_interval=1, enable=True):
```
I don't see any changes in this function. This should be Github's diff bug.

Stonesjtu commented on pytorch_memlab/utils.py:
```diff
@@ -1,5 +1,5 @@
 from calmsize import size as calmsize
-
+from math import isnan
```
Plz place this in the first line.

Stonesjtu commented on pytorch_memlab/utils.py:
```diff
@@ -1,5 +1,5 @@
 
 def readable_size(num_bytes):
-    return '{:.2f}'.format(calmsize(num_bytes))
+    return '' if isnan(num_bytes) else '{:.2f}'.format(calmsize(num_bytes))
```
I think it's ok to print NaN directly.

andyljones replied:
pandas uses `nan` as a filler value to indicate missing values; the HTML output is pretty ugly without this - all the empty lines with no reported data get nans.

Stonesjtu commented:
@andyljones I'm totally ok to drop python2.7 support. We have only < 1% python2.7 downloads.

Pandas dependency should not be a problem for me, it's a pretty common package. But I'm wondering if we can bypass pandas for the command-line interface.

andyljones commented on pytorch_memlab/line_profiler.py:
```diff
@@ -58,124 +60,165 @@ def add_function(self, func):
         """
         try:
-            code = func.__code__
+            codehash = hash(func.__code__)
```
pandas requires its index type be orderable; `code` doesn't have `__lt__` defined, while the hash does.

Stonesjtu replied:
Then can you add a comment line for such pandas requirement?

Stonesjtu replied:
BTW I prefer the naming schema `code_hash`.

andyljones commented on pytorch_memlab/line_profiler.py:
```diff
@@ -58,124 +60,165 @@ def add_function(self, func):
+                'func': func, 
+                'first_line': first_line,
+                'prev_line': first_line, 
+                'prev_record': -1}
```
When profiling an function that's calling another function that's being profiled, the `reset_memory_stat`s that happen in the called function mean that the peak/accumulated memory stats in the first line after the call are wrong - they give the peak/accumulated since the last line of the inner function. To get the correct stats, it has to keep track of what the last record for the caller was, and then take maxes/sums of peaks/accumulated stats since then.

andyljones commented on pytorch_memlab/__init__.py:
```diff
@@ -1,6 +1,6 @@
 from .courtesy import Courtesy
 from .mem_reporter import MemReporter
-from .line_profiler import LineProfiler, profile, profile_every, set_target_gpu
+from .line_profiler import LineProfiler, profile, set_target_gpu
```
Yep - forgot to restore it.

andyljones commented on pytorch_memlab/extension.py:
```diff
@@ -23,14 +23,17 @@ class MemlabMagics(Magics):
+              metavar='COLS',
+              action='append',
+              default=[],
+              help="""Columns to display. Can be specified multiple times to profile multiple
```
Done.

andyljones commented on pytorch_memlab/line_profiler.py:
```diff
@@ -58,124 +60,165 @@ def add_function(self, func):
+        records.columns = pd.MultiIndex.from_tuples([c.split('.') for c in records.columns])
+        return records
+
+    def print_stats(self, columns=['active_bytes.all.peak', 'reserved_bytes.all.peak'], func=None, stream=None):
```
Done

andyljones commented on pytorch_memlab/line_profiler.py:
```diff
@@ -1,12 +1,14 @@
 import torch
 from .utils import readable_size
+from IPython.display import HTML, display
+from functools import wraps
```
Done

andyljones commented on pytorch_memlab/line_profiler.py:
```diff
@@ -58,124 +60,165 @@ def add_function(self, func):
-        func.cur_idx = 1
+                assert False
+
+    def _refined(self):
```
Done

andyljones commented on pytorch_memlab/line_profiler.py:
```diff
@@ -58,124 +60,165 @@ def add_function(self, func):
+                assert False
+
+    def _refined(self):
+        # The records are line-by-line, but the stats we want to report are over periods.
```
I like to docstring things relevant to the interface, and comment things relevant to the implementation.

Stonesjtu replied:
Make sense. But I still suggest making this function name more specific.

andyljones replied:
Have done, is `_refined_line_records` now.

andyljones commented on pytorch_memlab/line_profiler.py:
```diff
@@ -58,124 +60,165 @@ def add_function(self, func):
+        acc_raw, peak_raw = raw.loc[:, acc_mask].values, raw.loc[:, peak_mask].values
+        acc_refined, peak_refined = acc_raw.copy(), peak_raw.copy()
+
+        for i, r in enumerate(self._raw):
```
Done

andyljones commented on pytorch_memlab/line_profiler.py:
```diff
@@ -58,124 +60,165 @@ def add_function(self, func):
+                continue
+            if r['prev'] == i-1:
+                continue
+            acc_refined[i] = acc_raw[r['prev']+1:i+1].sum(0)
```
Done

andyljones commented on pytorch_memlab/line_profiler.py:
```diff
@@ -58,124 +60,165 @@ def add_function(self, func):
+
+    def print_stats(self, columns=['active_bytes.all.peak', 'reserved_bytes.all.peak'], func=None, stream=None):
+        if len(self._raw) == 0:
+            print('No data collected.')
```
Done

andyljones commented:
Ok, I've resolved lots of the smaller issues. Bigger ones:

  * Text output: IPython handles objects that could be _either_ text or HTML by looking for [specific methods on the object returned by the cell](https://ipython.readthedocs.io/en/stable/config/integrating.html). This means we can definitely do text output, but it'll require returning an object from `print_stats`, which means `print_stats` won't be able to take a stream.
    What might be better is to move all the rich output to a new `.render()` method which'll return an object, and `print_stats` can pipe text to whatever stream it's been passed.  That has the advantage of preserving the existing API.
  * Terminal pandas: I'm inclined to use pandas to format the text output too. That's what I used for the text styling at the top of the thread:
![image](https://user-images.githubusercontent.com/662606/79637157-d2243880-8174-11ea-9e13-0cff94acd083.png)
  and I'm loathe to re-implement a pandas-free code path for terminal use.

Stonesjtu commented on pytorch_memlab/line_profiler.py:
```diff
@@ -58,124 +62,177 @@ def add_function(self, func):
+        else:
+            stream.write(html)
+
+    def print_func_stats(self, stats, func, **kwargs):
```
What's `stats` used for ?

andyljones replied:
Mistake.

Stonesjtu commented:
Let's clean things up based on ipython and pandas

andyljones commented:
Great. I'll do a full pass. Rather than list off individual names you'd like changed, is a there a style guide you can point me at?

Stonesjtu commented:
We followed Google's python style guide.

andyljones commented on pytorch_memlab/line_profiler.py:
```diff
@@ -58,124 +62,177 @@ def add_function(self, func):
+                style = style.bar([c], color=COLORS[i % len(COLORS)], width=99, vmin=0, vmax=maxes[c])
+            chunk = (style
+                        .format({c: readable_size for c in bytecols})
+                        .set_properties(subset=['code'], **{'text-align': 'left', 'white-space': 'pre', 'font-family': 'monospace'})
```
I don't think so? One applies to the code column, one to the table headers. Could pull them out to a big dict at the top of the file, but generally I don't like to split where-config-is-defined from where-config-is-used unless it's customizable.

willprice commented on pytorch_memlab/extension.py:
```diff
@@ -45,6 +48,10 @@ class MemlabMagics(Magics):
               nargs='*',
               default=None,
               help='Code to run under profiler. You can omit this in cell magic mode.')
+    @argument('-T',
```
Nope, it's there on master: https://github.com/Stonesjtu/pytorch_memlab/blob/master/pytorch_memlab/extension.py#L30

andyljones commented:
Right: 
* Now does text or html output
* Added docstrings
* Added a global interpreter reset cause @profile accumulates in an ugly way in an interpreter session
* Got the tests working 
* Got travis working
* Updated the readme
* Did lots of small fixes

Stonesjtu commented on pytorch_memlab/line_profiler.py:
```diff
@@ -1,26 +1,171 @@
+
+            string[qual_name] = merged.to_string(index=False)
+
+        return '\n\n\n'.join(['## {q}\n\n{c}'.format(q=q, c=c) for q, c in string.items()])
```
Better to move a `\n` from `'\n\n\n` after `{c}` to make a line break.

andyljones replied:
? Doing that would leave the text with an empty line at the bottom of the output.

Stonesjtu commented on pytorch_memlab/line_profiler.py:
```diff
@@ -1,26 +1,171 @@
+
+    def __repr__(self):
+        """Renders the stats as text"""
+        if len(self._line_records) == 0:
```
Can we use `if not self._line_records` here

Stonesjtu commented on pytorch_memlab/line_profiler.py:
```diff
@@ -1,26 +1,171 @@
+    def __repr__(self):
+        """Renders the stats as text"""
+        if len(self._line_records) == 0:
+            return 'No data collected'
```
Please add a line break here.

Stonesjtu commented on pytorch_memlab/line_profiler.py:
```diff
@@ -58,128 +204,143 @@ def add_function(self, func):
-                trace_stat=self.code_map[code],
-                stream=stream
-            )
+        line_records = _line_records(self._raw_line_records, self._code_infos)
```
Can we call `_line_records` in `_subset_line_records` to hide the implementation details.

andyljones replied:
Not without passing raw_line_records and code_infos to it too, and so confusing intent.

Stonesjtu commented on pytorch_memlab/line_profiler.py:
```diff
@@ -58,128 +204,143 @@ def add_function(self, func):
+        """ 
+        return RecordsDisplay(self.line_records(func, columns), self._code_infos)
 
+    def print_stats(self, func=None, columns=DEFAULT_COLUMNS, stream=sys.stdout):
```
I think we can seamlessly merge `print_stats` and `print_func_stats`.

andyljones replied:
done

Stonesjtu commented on pytorch_memlab/line_profiler.py:
```diff
@@ -58,128 +204,143 @@ def add_function(self, func):
-    function
-    The CUDA memory is collected only on the **current** cuda device.
+global_line_profiler = None
+def reset_global_line_profiler():
```
Semantically this function should modify the `global_line_profiler` **inplace**. We can add an `clear` method in LineProfiler instead.

andyljones replied:
done

Stonesjtu commented on pytorch_memlab/line_profiler.py:
```diff
@@ -1,26 +1,171 @@
+
+        string = {}
+        for qual_name, merged in self._line_records_merged_with_code().items():
+            left_align = '{{:{maxlen}s}}'.format(maxlen=max(map(len, merged.code)))
```
1. Plz use list-comprehension instead of `map`.
2. You can make it two lines for readability. (my personal taste)

ps. this *left-align* implementation looks weird as you commented.

andyljones replied:
done

andyljones commented on README.md:
```diff
@@ -52,43 +52,51 @@ the memory usage info for each line of code in the specified function/method.
-@profile
-def work():
+from pytorch_memlab import LineProfiler
+
```
The @profile decorator isn't able to return the display object, so it can't do HTML output. 

I've added an example to demo.ipynb.

andyljones commented on pytorch_memlab/extension.py:
```diff
@@ -72,15 +84,15 @@ def mlrun(self, line=None, cell=None, local_ns=None):
             return profiler
+        else:
+            defaults = [] if args.no_default_columns else DEFAULT_COLUMNS
+            return profiler.display(columns=defaults + args.column)
```
Most of the time you run the profiler, you're interested in peak allocation. Specifying `-c active_bytes.all.peak` every time is a pain. Problem is that if I put it in the default argument list for `column`, then - because it uses an append action - you wouldn't be able to override them the rare time that you don't actually want those columns at all.

andyljones commented on requirements.txt:
```diff
@@ -2,3 +2,4 @@ calmsize
 torch
```
It's 1.4; added it.

andyljones commented on setup.py:
```diff
@@ -26,9 +26,11 @@
     install_requires=[
         'setuptools',
         'calmsize',
+        'pandas>=0.18'
```
done

andyljones commented:
> can you make the function name an action rather than a thing? e.g. _subset_line_records to _extract_line_records, _line_records_merged_with_code to merge_line_records_with_code.

I've done this, but as an aside this is an idiom from functional programming. If a function has no side-effects, it effectively _is_ its return value, so it should be called a noun rather than a verb. Think of it as soft command-query separation.

OK. I've fulfilled most of the revision requests, and at this point the cleanup for the PR has taken substantially longer than implementing the original functionality. I respect that you've got high standards for the code in your projects, and I know you've invested time too in reviewing my code, but I'm afraid I can't put any more time towards this. If you're not happy with the PR as-is, that's okay: I'll instead publish the functionality I wrote as it's own package. I'd be disappointed, but unfortunately that's just how things work out sometimes.

Stonesjtu commented:
Andy, thanks for your valued time to make this PR and address my suggestions. This PR has been impressive since the first proposal. I'm just trying to help you make it more *pythonic* and readable (for most python coders). It's a common situation that people spend more time addressing code reviews than functionality.

I can merge this PR first and then do some changes myself, but I think it's better for you to make this PR a complete one. Sadly you don't have that amount of time.

andyljones commented:
Huurah! Thanks very much
